File: src/stores/whois.ts
import { defineStore } from 'pinia';
import { ref } from 'vue';
import { invoke } from '@tauri-apps/api/core';
import type { WhoisInfo } from '../models/whois';

export const useWhoisStore = defineStore('whois', () => {
  const whoisInfo = ref<WhoisInfo | null>(null);
  const loading = ref<boolean>(false);
  const error = ref<string | null>(null);

  const fetchWhois = async (domain: string) => {
    loading.value = true;
    error.value = null;

    try {
      const result = await invoke<WhoisInfo>('lookup_whois', { domain });
      whoisInfo.value = result;
    } catch (e) {
      error.value = e as string;
      console.error('Failed to fetch WHOIS:', e);
    } finally {
      loading.value = false;
    }
  };

  const clear = () => {
    whoisInfo.value = null;
    error.value = null;
  };

  return {
    whoisInfo,
    loading,
    error,
    fetchWhois,
    clear,
  };
});

File: src-tauri/src/commands/whois.rs
use crate::adapters::whois::WhoisAdapter;
use crate::models::whois::WhoisInfo;
use tauri::AppHandle;

#[tauri::command]
pub async fn lookup_whois(app_handle: AppHandle, domain: String) -> Result<WhoisInfo, String> {
    let adapter = WhoisAdapter::with_app_handle(app_handle);
    adapter.lookup(&domain).await
}
